README.md
Shopify API access for Meteor.

## Simple, Convenient, Synchronous
```javascript
// Server
var api = new Shopify.API({
    shop: "your-shop",
    access_token: "Your access token"
});

var orderCount = api.countOrders(); // => 1022

var recentOpenOrders = api.getOrders();
console.log(recentOpenOrders.length); // => 250
console.log(recentOpenOrders[0]); // => { name: "#2022", ... }

// Combine countOrders and getOrders:
var allOpenOrders = api.getAllOrders({ status: "any" });
console.log(allOpenOrders.length); // => 1022

var order102420 = api.getOrder({ id: 102420 });
console.log(order102420); // => { name: "#2022", ... }
```

## Public & Private Apps
This package supports both Public and Private Apps.

### Private Apps
If you just want to access your own shop's data, then it's easiest to make a Private App.  Login to your Shopify Partner account and create it, and take note of your `api_key` and `password`.

```javascript
// Private App API access (server)
ShopifyAPI = new Shopify.API({
    shop: "your-shop",
    api_key: "Your API Key",
    password: "Your API Password"
});
```

### Public Apps
If you need to access multiple shops, then you need a Public App.  The first time a user uses your app, they will need to authenticate with OAuth 2.  But the access code returned from Shopify can be turned into a permanent `access_token`.  Login to your Shopify Partner account and create your Public App, and take note of your `api_key` and `secret`.  You will also need to whitelist whichever `redirect_uri` you choose to use (App -> Your App -> Edit app settings -> Redirect URL).  If you use the `Shopify.PublicAppOAuthAuthenticator`, then whitelist `localhost:3000/__shopify-auth` (with 2 underscores).  And be sure to replace the `localhost:3000` when you go into production!

```javascript
// First-time Public App API access

// On the server (since you don't want to expose your secret to the client)
Shopify.addKeyset("default", {
    api_key: "Your API Key",
    secret: "Your API Shared Secret",
});

// On the client
var authenticator = new Shopify.PublicAppOAuthAuthenticator({
    shop: "your-shop",
    api_key: "Your API Key",
    keyset: "default",
    scopes: "all",

    // Once the login succeeds and a permanent access token is acquired, store
    // it in the user's profile for future use.
    onAuth: function(accessToken) {
        Meteor.users.update(Meteor.userId(), {
            $set: { "profile.shopifyAccessToken": access_token }
        });
    }
});

authenticater.openAuthTab();
```

NOTE: in the above scenario, the `access_token` is leaked to the client.  If this is a concern to you, then:

```javascript
// server/init.js
Shopify.harden(); // do not send accessToken to client
Shopify.onAuth(function(access_token, authConfig, userId) {
    // You can store the access_token in the user object for future use.  Don't
    // store it in the profile or it will leak to the client.
    Meteor.users.update(userId, {
        $set: {
            shop: authConfig.shop,
            access_token: access_token,
        },
    });
});
```

`scopes` (Public Apps only) tell Shopify what access your app needs.  The user will be prompted to grant access to the scopes your App requests when he/she navigates to `authURL`.  The default is `"all"`, but a comma separated string like `"read_orders,read_products"` should be given to the authenticator if you know you don't need access to everything.  The supported scopes are:

* `"read_content"`
* `"write_content"`
* `"read_themes"`
* `"write_themes"`
* `"read_products"`
* `"write_products"`
* `"read_customers"`
* `"write_customers"`
* `"read_orders"`
* `"write_orders"`
* `"read_script_tags"`
* `"write_script_tags"`
* `"read_fulfillments"`
* `"write_fulfillments"`
* `"read_shipping"`
* `"write_shipping"`

See below for more OAuth options, like if you want to authenticate in an iframe or with your own UI.

#### Signature verification
As of 1.6.0, if you use `Shopify.PublicAppOAuthAuthenticator`, then the hmac sent by Shopify is verified during auth.  This uses the shared secret stored in your keyset.

If the signature is invalid, an error message is shown: Shopify Auth failed because of invalid signature/shared secret.

This behavior can be disabled by calling `Shopify.ignoreHMACs()`.

```javascript
// Second-time Public App API access

// On the server or client.
var = new Shopify.API({
    shop: "your-shop",
    access_token: "Your access token"
});
```

You can also use a keyset if you don't want the client to see the `access_token`.  In this case, be sure to call `Shopify.harden()` so that the client will not receive `access_token` as an `onAuth` parameter in `PublicAppOAuthAuthenticator`.

```javascript
// On the server, e.g. server/init.js
Shopify.harden();

// On the client
var authenticator = new Shopify.PublicAppOAuthAuthenticator({
    ...
    onAuth: function(err, access_token) {
        console.assert(typeof access_token === "undefined");
    },
});

```

Instead you can save `access_token` and create a keyset using a global `onAuth` handler:

```javascript
Shopify.onAuth(function(access_token, authenticatorConfig, userId) {
    ShopifyCredentials.upsert(userId, {
        shop: authenticatorConfig.shop,
        api_key: authenticatorConfig.api_key,
        access_token: access_token,
    });
});
```

### Keysets
You don't want to expose your `secret` (Public Apps), or `password` (Private Apps) to the client.  Therefore, use a `keyset` instead.

```javascript
// Server
Meteor.startup(function() {
    // For Public Apps
    Shopify.addKeyset("default", {
        api_key: "Your API Key",
        secret: "Your API Shared Secret"
    });

    // Or for Private Apps
    Shopify.addKeyset("default", {
        api_key: "Your API Key",
        password: "Your API Password"
    });

    // Or for Public Apps when the user has logged in
    Shopify.addKeyset("default", {
        access_token: "This shop's OAuth permanent access token"
    });
});

// Client (or Server)
var api = new Shopify.API({
    shop: "your-store",
    keyset: "default"
});
```

You can also delete a keyset when you're done with it:
```javascript
// Server
Shopify.removeKeyset("default");
```

And you can check if a keyset exists by name:
```javascript
// Server
Shopify.keysetExists(shopName); // => true or false
```

### Insecure Client
If you are control of the browser environment, you can pass your `api_key` and `password/secret` in `options`, ignoring `options.keyset`.  Be sure to set `options.insecure` to `true`, and use underscored method names like `API._getOrders` in order to make the calls via AJAX.

If you call non-underscored methods, like `API.getOrders` on the client, requests will be proxied through the server using the Meteor Method `"froatsnook:shopify/call"`.

## Embedded App SDK
Version 1.3.0 adds experimental Embedded App SDK support.  The current workflow is:

* Set `embedded_app_sdk` to `true` in your `Shopify.PublicAppOAuthAuthenticator`
* Call `Shopify.getEmbeddedAppAPI` to load Shopify's `ShopifyApp` object
* Initialize it by providing an api key and the store's domain
* Make sure you've enabled embedded settings in your partner account (Apps -> Your App -> Edit app settings -> Embedded settings)
* Make sure the `redirect_uri` you're using (defaults to `Meteor.absoluteUrl("/__shopify-auth")`) is set in your partner account (Apps -> Your App -> Edit app settings -> Redirection URL)
* Do your thing

See an example [here](examples/EmbeddedAppSDKTest).

For a list of `ShopifyApp` methods, see Shopify's documentation [here](https://docs.shopify.com/embedded-app-sdk/methods).

If you have problems, questions, or suggestions, please make an issue on github.

### Shopify.getEmbeddedAppAPI

This function should be called on the client from within the iframe which is
embedded in the store in order to retrieve the Embedded App SDK `ShopifyApp`
object.  You can use this object to show messages and do other Embedded App SDK
stuff.

```javascript
Shopify.getEmbeddedAppAPI(function(err, ShopifyApp) {
    if (err) {
        toastr.error("Failed to load ShopifyApp");
        return;
    }

    ShopifyApp.init({
        apiKey: "...",
        shopOrigin: "https://" + shop + ".myshopify.com"
    });

    ShopifyApp.flashNotice("It works");
});
```

## Client and Server APIs
While the Server API is synchronous using `Fibers`, the client API is not.  So each method needs an additional callback, with `err` and possible return value.

```javascript
// Server
var count = api.countOrders();

// Client
api.countOrders(function(err, count) {
    ...
});

```

Instead of `err`, `Error`s are thrown on the server.

If you want to use the asynchronous API on the server, for example to share code between (an insecure) client and server, use underscored methods, like `api._getOrders`.

## API Updates
If the Shopify API changes and something breaks, of course please create an issue.  However, `API.define` and `API.defineConcat` are provided to hold you over.

Just call:

```javascript
Shopify.API.define({
    "name": "getProductMetafieldByID",
    "method": "GET",
    "path": "/admin/products/#{productId}/metafields/#{id}.json",
    "returns": "metafield",
    "description": "Get a single product metafield by its ID"
});
```

to directly modify `Shopify.API.prototype`, adding the `getProductMetafieldByID` method (sync on server, async on client), as well as `_getProductMetafieldByID` (async on both).  Since `#{id}` is in `path`, it will take a required param `id`.  The optional `returns` indicates that Shopify returns json like `{ "metafield": { ... } }` and `getProductMetafieldByID` should return the `metafield` of that object.

Sometimes Shopify specifies their API with two of the same params, such as `GET /admin/orders/#{id}/transactions/#{id}.json`.  Be sure to rename one of these, like `GET /admin/orders/#{orderId}/transactions/#{id}.json`.

Defining `getAllXYZ` methods is possible in the same way using:

```javascript
Shopify.API.defineConcat({
    "name": "getAllOrders",
    "count": "countOrders",
    "fetch": "getOrders"
});
```

## Get All
The Shopify API limits some calls, such as `getOrders`, to 250 items.  Get All methods are also defined for such methods in order to automate the work of counting the collection, fetching the pages individually, and then concatenating the results.

```javascript
var allOpenOrders = api.getAllOrders();
var allOrders = api.getAllOrders({ status: "any" });
```

The currently available Get All methods are:
* `getAllOrders`
* `getAllProducts`
* `getAllCollects`
* `getAllCustomers`

## Rate Limiting
This library does its best to not exceed the Shopify rate limit of one call each half second with bursts of up to 40 calls.

With each reply, Shopify returns a `X-Shopify-Shopify` header like `"32/40"`.  Once this reaches the `api.config.backoff` (default `35`), API calls will be queued and fired each half second until again below the backoff limit.

## Custom OAuth
`PublicAppOAuthAuthenticator.openAuthTab` is provided as the simplest way to do OAuth.  However, you can also take control at various steps in the process if you want a more customized experience.

For new apps, you will need to whitelist the `redirect_uri` you're using.  This defaults to `Meteor.absoluteUrl("/__shopify-auth")` if you're using the `Shopify.PublicAppOAuthAuthenticator`.  The setting can be found in your partner account (Apps -> Your App -> Edit app settings -> Redirection URL).

### All scenarios
```javascript
// On the server (since you don't want clients to know your API Shared Secret).
Shopify.addKeyset("default", {
    api_key: "Your Shopify App's API Key",
    secret: "Your Shopify App's API Shared Secret"
});
```

### Scenario 1: Automatic
```javascript
     console.assert(Meteor.isClient);
     var authenticator = new Shopfiy.PublicAppOAuthAuthenticator({
         shop: "my-shop-40",
         api_key: "Your Shopify App's API Key",
         keyset: "default",
         scopes: "read_products,read_customers",
         onAuth: function(access_token) {
             Meteor.call("SaveShopifyCredentials", access_token);

             var api = new Shopify.API({ authenticator: authenticator });
             api.getProducts(function() { ... });
         }
     });

     // Open a new tab in the browser where the user can login with user name and password.  The user will then see the scopes that you requested and will be able to click an "Install App" button.  Shopify will redirect the tab back to this Meteor app, which will then exchange the given code for an access_token, and close the tab.
     authenticator.openAuthTab();
```

### Scenario 2: iframe
```javascript
     console.assert(Meteor.isClient);

     // Let the user install your app in an iframe instead of a new tab.
     var iframe = $("<iframe>");
     iframe.appendTo("body");

     var authenticator = new Shopify.PublicAppOAuthAuthenticator({
         shop: "my-shop-40",
         api_key: "Your Shopify App's API Key",
         keyset: "default",
         scopes: "read_products,read_customers",

         // Navigate iframe to a success message after authentication
         post_auth_uri: Meteor.absoluteUrl("shopify-login-success"),

         // After authentication, save the access_token, then remove the
         // iframe after 2 seconds.
         onAuth: function(access_token) {
             Meteor.call("SaveShopifyCredentials", access_token);

             Meteor.setTimeout(function() {
                 iframe.remove();
             }, 2000);

             var api = new Shopify.API({ authenticator: authenticator });
             api.getProducts(function() { ... });
         }
     });

     // Navigate the iframe to a Shopify page where the user can login with user name and password.  The user will then see the scopes that you requested and will be able to click an "Install App" button.  Shopify will redirect the tab back to this Meteor app, which will then exchange the given code for an access_token, and close the tab.
     iframe.attr("src", authenticator.authURL);
```

### Scenario 3: direct navigation
```javascript
     console.assert(Meteor.isClient);

     var authenticator = new Shopify.PublicAppOAuthAuthenticator({
         shop: "my-shop-40",
         api_key: "Your Shopify App's API Key",
         keyset: "default",
         scopes: "read_products,read_customers",

         // Logging in will redirect here.
         redirect_uri: Meteor.absoluteUrl("shopify-logging-in"),

         // Navigate iframe to a success message after authentication
         post_auth_uri: Meteor.absoluteUrl("shopify-login-success"),

         // This will be called after authenticator.getPermanentAccessToken
         // succeeds.
         onAuth: function(access_token) {
             Meteor.call("SaveShopifyCredentials", access_token);

             var api = new Shopify.API({ authenticator: authenticator });
             api.getProducts(function() { ... });
         }
     });

     // Start login by navigating.
     location.href = authenticator.auth_uri;

     // ... acquire code from redirect_uri using WebApp.connectHandlers.use or an iron:router route.

     // Upgrade code to permanent access_token
     authenticator.getPermanentAccessToken(code);
```

## Testing

Tests can be run using:

`meteor test-packages ./`

This starts a Shopify API Simulator and runs tests, which can be checked at `http://localhost:3000/`.

## API

### Shopify.API constructor

Shopify API access objects can be created with `new Shopify.API`:

```javascript
var api = new Shopify.API({
    shop: "test-shop-10",
    backoff: 38,
    insecure: true,
    additionalHeaders: { "X-DEV-MOOD": "Pensive" }
});
```

The available options are:

* `options.shop {String}` The target shop name, like `"my-store"` in `"my-store.my-shopify.com"`.

CHANGELOG.md
1.6.1 - 2015-12-19
==================
* Allow user to explicitly specify scopes as "all"
* Add BasicExample

1.6.0 - 2015-12-19
==================
* The `hmac` returned by Shopify during auth is now checked (`Shopify.PublicAppOAuthAuthenticator` only).
* Added Shopify.ignoreHMACs to disable this check.

1.5.1 - 2015-12-17
==================
* Fix private API keyset use thanks to [toszter](https://github.com/toszter)

1.5.0 - 2015-11-11
==================
* Add server-side `Shopify.onAuth`
* Add `Shopify.harden()` to prevent leaking `access_token` to client
* Add `Shopify.keysetExists`

1.4.4 - 2015-11-04
==================
* Fix `_ undefined` error in getAllCustomers and friends.

1.4.3 - 2015-11-04
==================
* Another auth fix, this time thanks to [sum1youno](https://github.com/sum1youno)

1.4.2 - 2015-10-11
==================
* Fix auth

1.4.0 - 2015-07-06
==================
* Fix `addKeyset` so that providing `access_token` "overrules" other options
* Further improvements to embedded app SDK example
* Fix automatic Authenticator creation when keyset given as option in `new Shopify.API`
* Fix POST requests when there's a request body

1.3.1 - 2015-05-26
==================
* Improve embedded app SDK example

1.3.0 - 2015-05-24
==================
* Add experimental Embedded App SDK support

1.2.0 - 2015-04-22
==================
* Add `getAllCustomers`

1.1.0 - 2015-04-16
==================
* Add `options.debug` to `Shopify.API` constructor

1.0.0 - 2015-03-18
==================
* Major client security overhaul

0.1.0 - 2015-03-17
==================
* Initial version



lib/api.js
// Create an API access object.
//
// @example
//     var api = new Shopify.API({
//         shop: "test-shop-10",
//         backoff: 38,
//         insecure: true,
//         additionalHeaders: { "X-DEV-MOOD": "Pensive" }
//     });
//
// @param options {Object}
// @param options.shop {String} The target shop name, like `"my-store"` in `"my-store.my-shopify.com"`.
// @param [options.keyset] {String} Name of keyset (which must be registered on the server via `Shopify.addKeyset`).  A keyset can be used instead of providing `api_key+password` or `api_key+secret` or `access_token`.
// @param [options.api_key] {String} Your App's API Key (Public and Private Apps).  This can be provided on the server (instead of a keyset), and on the client.
// @param [options.password] {String} Your App's API Password (Private Apps only).  This can be provided on the server (instead of a keyset), and on the client if `options.insecure` is `true`.
// @param [options.secret] {String} Your App's API Shared Secret (Public Apps only).  This can be provided on the server (instead of a keyset), and on the client if `options.insecure` is `true`.
// @param [options.access_token] {String} This shop's permanent access token acquired via OAuth2.  This can be provided on the server (instead of a keyset), and on the client if `options.insecure` is `true`.
// @param [options.backoff] {Number} (default=`35`) The number of requests in Shopify's call bucket before requests are enqueued and fired off every half second.
// @param [options.insecure] {Boolean} (default=`false`) If `true`, then various security assertions and warnings will be ignored.
// @param [options.additionalHeaders] {Object} (default=`{ }`) Headers here will be sent with each call to Shopify (mainly for testing).
// @param [options.debug] {Boolean} If true, then `console.log` debug info.
var API = Shopify.API = function ShopifyAPI(options) {
    check(options, Object);
    check(options.shop, Shopify._IsNonEmptyString);

    // Until authenticated, or while waiting for the API rate limit to cool
    // down, requests are enqueued.
    //
    // [{ method, uri, [data], callback }, ...]
    this._requestQueue = [];

    this.config = {
        shop: options.shop,
        backoff: options.backoff || 35,
        insecure: options.insecure || false,
        additionalHeaders: options.additionalHeaders || { },
        _HOST_OVERRIDE: options._HOST_OVERRIDE || null,

        keyset: options.keyset || null,
        api_key: options.api_key || null,
        password: options.password || null,
        secret: options.secret || null,
        access_token: options.access_token || null,
        debug: options.debug || false
    };

    // On the server we can immediately replace the keyset with the credentials.
    if (Meteor.isServer && this.config.keyset) {
        var keyset = Shopify._keysets[this.config.keyset];
        switch (keyset.type) {
            case "private":
                this.config.api_key = keyset.api_key;
                this.config.password = keyset.password;
                break;
            case "oauth":
                this.config.api_key = keyset.api_key;
                this.config.secret = keyset.secret;
                break;
            case "public":
                this.config.access_token = keyset.access_token;
                break;
        }

        this.config.keyset = null;
    }

    // If user has an `access_token` or `password`, an authenticator can be
    // created.  Otherwise the user must later provide an authenticator with
    // API.addAuthenticator.
    this._authenticator = null;

    if (Meteor.isClient) {
        if (options.password && !options.insecure) {
            console.warn("Don't expose your App's password on the client!  Use `options.keyset` instead and `Shopify.addKeyset` on the server.  Use `options.insecure=true` to disable this warning.");
        }

        if (options.secret && !options.insecure) {
            console.warn("Don't expose your App's shared secret on the client!  Use `options.keyset` instead and `Shopify.addKeyset` on the server.  Use `options.insecure=true` to disable this warning.");
        }
    }

    // Try to make an authenticator even if client.
    if (this.config.password) {
        if (!this.config.api_key) {
            throw new Error("api_key is required for Private Apps (assuming Private App because you provided a password)");
        }

        var authenticator = new Shopify.PrivateAppAuthenticator({
            api_key: this.config.api_key,
            password: this.config.password,
        });

        this.addAuthenticator(authenticator);
    } else if (this.config.access_token) {
        var authenticator = new Shopify.PublicAppAuthenticator({
            access_token: this.config.access_token,
        });

        this.addAuthenticator(authenticator);
    }

    // From X-Shopify-Shop-Api-Call-Limit, like 32 in "32/40".
    this._lastCallLimit = 0;

    // Number of requests currently waiting.  If at backoff, then will
    // incremenet queue, wait a half second for each item in the queue, then
    // decrement the queue and make the API call.
    this._queueLength = 0;

    // Fix for 1.4.4: in functions created with `new Function`, underscore is
    // unavailable.  Since these functions have `API` as `this`, they can
    // access it here.
    this.underscore = _;
};

// Add an authenticator.
API.prototype.addAuthenticator = function(authenticator) {
    this._authenticator = authenticator;
};

Object.defineProperty(API.prototype, "authenticated", {
    get: function() {
        return !!this._authenticator;
    }
});

// Wait (if necessary) for rate limiting purposes, then call callback.
API.prototype._waitAsync = function(callback) {
    var self = this;

    // If not yet at backoff, then no wait is necessary.
    if (this._lastCallLimit < this.config.backoff) {
        Meteor.setTimeout(callback, 0);
        return;
    }

    this._queueLength++;
    if (this.config.debug) { console.log("Request will wait %sms because of call limit", 500*this._queueLength); }
    Meteor.setTimeout(function() {
        self._queueLength--;
        callback();
    }, 500*this._queueLength);
};

// When the client does API.getOrders(), this request has to be proxied to the
// server for security reasons (cross-domain AJAX not allowed and to prevent
// API Secret exposure).  The proxying happens by calling this Meteor Method.
//
// Server-only because we don't want the client to try to simulate it.
if (Meteor.isServer) {
    Meteor.methods({
        // @param name {String} Name of the meteor method.
        // @param options {Object} Method options.
        // @param config {Object} API.config object (for finding/created cached API object).
        "froatsnook:shopify/call": function(name, options, config) {
            // Retrieve cached server-side API object.
            var cachedAPI = Shopify._getCachedAPI(config);

            // Call the synchronous version of the method.
            return cachedAPI[name](options);
        }
    });
}

// Methods.

// Define an API method.
//
// @param info {Object} Endpoint info
// @param info.name {String}
// @param info.method {String} HTTP method of API endpoint, like "GET"
// @param info.path {String} Path of API endpoint, like "/admin/orders.json"
// @param [info.description] {String}
API.define = function(info) {
    check(info.name, Shopify._IsNonEmptyString);
    check(info.method, Shopify._IsNonEmptyString);
    check(info.path, Shopify._IsNonEmptyString);

    var description = info.method + " \"" + info.path + "\"";
    if (info.description) {
        description += ": " + info.description;
    }

    var paramFinder = /#{(.+?)}/g;
    var params = [];

    var m;
    while ((m = paramFinder.exec(info.path))) {
        params.push(m[1]);
    }

    // If params are given (e.g. `id` in GET /admin/orders/:id.json), then we
    // have to substitute `options[param]` into `path` for each param.
    var paramCode;
    if (params.length > 0) {
        var lines = [];
        params.forEach(function(param) {
            lines.push("");
            lines.push("    if (!options['" + param + "']) {");
            lines.push("        throw new Error('Missing option " + param + "');");
            lines.push("    }");
            lines.push("");
            lines.push("    path = path.replace('#{" + param + "}', options['" + param + "']);");
        });

        // Make helpers for
        lines.push("var paramCount = " + params.length + ";");
        lines.push("var isParam = { '" + params.join("': true, '") + "': true };");

        paramCode = lines.join("\n");
    } else {
        paramCode = "var paramCount = 0, isParam = { };";
    }

    var returnCode;
    if (info.returns) {
        returnCode = "            callback(null, res.data." + info.returns + ");\n";
    } else {
        returnCode = "            callback(null, res.data);\n";
    }

    API.prototype["_" + info.name] = new Function([
        // Name the function
        "return function ",
        info.name,
        // Takes options and callback.
        "(options, callback) {\n",
        // expose underscore
        "    var _ = this.underscore;\n",
        // Add the description in a comment for toString
        "    /* " + description + " */\n",
        "    var self = this;\n",

        "\n",
        "    if (typeof options === 'function') {\n",
        "        callback = options;\n",
        "        options = { };\n",
        "    }\n",
        "\n",

        "    this._waitAsync(function() { // for rate limiting\n",

        "    var method = '" + info.method + "';\n",
        "    var path = '" + info.path + "';\n",

        "\n",
        paramCode,
        "\n",

        // Extract request body content from `content' or `data'.
        "    var req = { };\n",
        "    if (options.content) {\n",
        "        req.content = options.content;\n",
        "        delete options.content;\n",
        "    }\n",
        "    if (options.data) {\n",
        "        req.data = options.data;\n",
        "        delete options.data;\n",
        "    }\n",
        "\n",

        // Using paramCode helpers from above, generate a query string from all
        // the other options.
        "    var queryParams = [];\n",
        "    for (var option in options) {\n",
        "        if (isParam[option]) {\n",
        "            continue;\n",
        "        }\n",
        "\n",
        "        var optionVal = options[option];\n",
        "        if (optionVal instanceof Date) {\n",
        "            optionVal = optionVal.toISOString().slice(0, 16).replace('T', ' ');\n",
        "        }\n",
        "\n",
        "        queryParams.push(option + '=' + optionVal);\n",
        "    }\n",
        "\n",
        "    var search = queryParams.length === 0 ? '' : '?' + queryParams.join('&');\n",
        "\n",

        "    var url = 'https://' + self.config.shop + '.myshopify.com' + path + search;\n",
        "    self._authenticator.authenticate(req);\n",
        "    var res = self._httpCall(method, url, req, function(err, res) {\n",
        "        if (err) { return callback(err); }\n",
        "\n",
        "        // Try again if hit rate limit\n",
        "        if (res.statusCode === 429) { return " + info.name + "(options, callback); }\n",
        "\n",
        "        // Set call limit\n",
        "        var callLimit = res.headers['x-shopify-shop-api-call-limit'];\n",
        "        if (callLimit) {\n",
        "            var callLimitParts = callLimit.split('/');\n",
        "            var bucket = parseInt(callLimitParts[0], 10);\n",
        "            if (isNaN(bucket)) {\n",
        "                console.warn('Invalid call limit %s', callLimit);\n",
        "            } else {\n",
        "                self._lastCallLimit = parseInt(bucket);\n",
        "            }\n",
        "        }\n",
        "\n",
        "        // 2XX codes indicate success\n",
        "        if (200 <= res.statusCode && res.statusCode <= 299) {\n",
        returnCode,
        "            return;\n",
        "        }\n",
        "\n",
        "        // It failed\n",
        "        callback(res.statusCode);\n",
        "    });\n",

        "    }); // waitAsync\n",
        "}"
    ].join(""))();

    if (Meteor.isServer) {
        API.prototype[info.name] = Meteor.wrapAsync(API.prototype["_" + info.name]);
    } else {
        API.prototype[info.name] = function(options, callback) {
            if (typeof options === "function") {
                callback = options;
                options = { };
            }

            if (typeof callback !== "function") {
                throw new Error("Missing required param 'callback'");
            }

            Meteor.call("froatsnook:shopify/call", info.name, options, this.config, callback);
        };
    }
};

Shopify._APIMethods.forEach(function(apiMethod) {
    API.define(apiMethod);
});

API.defineConcat = function(info) {
    check(info.name, Shopify._IsNonEmptyString);
    check(info.count, Shopify._IsNonEmptyString);
    check(info.fetch, Shopify._IsNonEmptyString);

    // Orders need options.status set to "open" if getAllOrders.
    var addStatusCode;
    if (info.name === "getAllOrders") {
        addStatusCode = [
            "",
            "    // By default, getOrders only returns open orders.  So",
            "    // getAllOrders should do the same.  If not explicitly provided",
            "    // then getOrders and countOrders will behave differently.",
            "    if (!options.status) {",
            "        options.status = 'open';",
            "    }",
            ""
        ].join("\n");
    } else {
        addStatusCode = "\n";
    }

    API.prototype["_" + info.name] = new Function([
        "return function " + info.name + "(options, callback) {",
        // expose underscore
        "    var _ = this.underscore;\n",
        "    var self = this;",
        "    if (typeof options === 'function') {",
        "        callback = options;",
        "        options = { };",
        "    } else {",
        "        options = _.clone(options);",
        "    }",
        addStatusCode,

        "    // Check how many items to fetch total",
        "    if (self.config.debug) { console.log('" + info.name + ": counting via _" + info.count + "'); }",
        "    if (self.config.debug) { console.log('" + info.name + ": fetching via _" + info.fetch + "'); }",
        "    this._" + info.count + "(options, function(err, count) {",

        "        // Check for count fail",
        "        if (err) {",
        "            if (self.config.debug) { console.log('" + info.name + ": count error: %s', err); }",
        "            callback('count failed: ' + err);",
        "            return;",
        "        }",
        "",
        "        if (self.config.debug) { console.log('" + info.name + ": count=%s', count); }",
        "",
        "        // Request 250 at a time starting at page 1",
        "        options.limit = 250;",
        "        options.page = 1;",
        "",
        "        // Append results here",
        "        var results = new Array(count);",

        "        // Request a page each time caller is called",
        "        (function caller(n) {",
        "            if (self.config.debug) { console.log('" + info.name + ": caller(%s)', n); }",

        "            // Return if done",
        "            if (n >= count) {",
        "                if (self.config.debug) { console.log('" + info.name + ": success!'); }",
        "                callback(null, results);",
        "                return;",
        "            }",
        "",

        "            // Fetch a page",
        "            self._" + info.fetch + "(options, function(err, list) {",

        "                // Check for fetch fail",
        "                if (err) {",
        "                    if (self.config.debug) { console.log('" + info.name + ": fetch error: %s', list.length); }",
        "                    callback('fetch (page ' + options.page + ') failed: ' + err);",
        "                    return;",
        "                }",

        "                if (self.config.debug) { console.log('" + info.name + ": fetch list.length=%s', list.length); }",
        "",
        "                // Add to results",
        "                for (var i = 0, len = list.length; i < len; i++) {",
        "                    results[n + i] = list[i];",
        "                }",
        "",
        "                // Recurse",
        "                options.page++;",
        "                caller(n + options.limit);",

        "            });", // fetch

        "        })(0);", // caller
        "",

        "    });", // count

        "};" // new Function
    ].join("\n"))();

    if (Meteor.isServer) {
        API.prototype[info.name] = Meteor.wrapAsync(API.prototype["_" + info.name]);
    } else {
        API.prototype[info.name] = function(options, callback) {
            if (typeof options === "function") {
                callback = options;
                options = { };
            }

            Meteor.call("froatsnook:shopify/call", info.name, options, this.config, callback);
        };
    }
};

Shopify._APIConcatMethods.forEach(function(apiMethod) {
    API.defineConcat(apiMethod);
});

API.prototype._httpCall = function(method, uri, req, callback) {
    if (this.config._HOST_OVERRIDE) {
        var path;
        if (Meteor.isServer) {
            var url = Npm.require("url");
            var parsedUri = url.parse(uri);
            path = parsedUri.path;
        } else {
            var link = document.createElement("a");
            link.href = uri;
            path = link.pathname + link.search;
        }

        uri = this.config._HOST_OVERRIDE + path;
    }

    if (!req.headers) {
        req.headers = { };
    }

    for (var key in this.config.additionalHeaders) {
        var val = this.config.additionalHeaders[key];
        if (!req.headers[key]) {
            req.headers[key] = val;
        }
    }

    if (this.config.debug) { console.log("%s %s", method, uri); }

    HTTP.call(method, uri, req, callback);
};

